warehouse/query-core/src/main/java/datawave/next/scanner/QueryDataConsumer.java

```diff
@@ -0,0 +1,161 @@
+    public void run() {
+        try {
+            if (log.isDebugEnabled()) {
+                log.debug("query data consumer started");
+            }
+            while (iterator.hasNext()) {
+                QueryData queryData = iterator.next();
+                if (log.isDebugEnabled()) {
+                    log.debug("got query data: {}", queryData.getRanges().iterator().next().getStartKey().toStringNoTime());
+                }
+                if (queryData == null) {
+                    stats.incrementNullDataSeen();
+                    log.info("query data was null");
+                    continue;
+                }
+                if (queryData.getSettings() == null) {
+                    log.info("query data settings was null");
+                }
+                if (queryData.getRanges() == null) {
+                    log.info("query data ranges was null");
+                }
+                stats.incrementQueryDataSeen();
+
+                Preconditions.checkArgument(queryData.getSettings().size() == 1);
+                Preconditions.checkArgument(queryData.getRanges().size() == 1);
+
+                Range range = queryData.getRanges().iterator().next();
+                if (isDocumentRange(range)) {
+                    putDocId(queryData, range);
+                } else {
+                    putFiScan(queryData, range);
+                }
+            }
+        } catch (Exception e) {
+            log.error("ScannerChunkConsumer saw error", e);
+            throw new RuntimeException(e);
+        } finally {
+            executing.set(false);
+            if (log.isDebugEnabled()) {
+                log.debug("query data consumer stopped");
+            }
+
+            config.getStats().setConsumerStats(stats);
+        }
+    }
+
+    private boolean isDocumentRange(Range range) {
+        Key start = range.getStartKey();
+        return start.getColumnFamily().getLength() > 0;
+    }
+
+    private void putFiScan(QueryData queryData, Range range) throws TableNotFoundException {
+        stats.incrementNumShardScans();
+
+        IteratorSetting settings = queryData.getSettings().get(0);
+
+        IteratorSetting next = new IteratorSetting(settings.getPriority(), "DocIdQueryIterator", DocIdQueryIterator.class);
+        next.addOption(QueryOptions.QUERY, queryData.getQuery());
+        next.addOption(QueryOptions.START_TIME, settings.getOptions().get(QueryOptions.START_TIME));
+        next.addOption(QueryOptions.END_TIME, settings.getOptions().get(QueryOptions.END_TIME));
+        next.addOption(QueryOptions.INDEXED_FIELDS, settings.getOptions().get(QueryOptions.INDEXED_FIELDS));
+        if (settings.getOptions().containsKey(QueryOptions.DATATYPE_FILTER)) {
+            next.addOption(QueryOptions.DATATYPE_FILTER, settings.getOptions().get(QueryOptions.DATATYPE_FILTER));
+        }
+
+        Scanner scanner = config.getClient().createScanner(queryData.getTableName(), config.getAuthorizations());
```

Comment:
shift this to our friend the scannerfactory

Reply:
Adding a TODO..